Migeran LibGodot Feature

kisg opened the pull request:
## Features

* Compile Godot Engine as either static or shared library.
* Expose Godot Engine controls (startup, iteration, shutdown) over the GDExtension API to a host process.
* On Apple Platforms, Godot Engine windows (both the main window, and other windows created after startup) may be rendered into surfaces provided by the host process.
  * This support can be extended to other platforms, read on for details.
* Sample applications (separate repository)
  * C++ Sample (tested on MacOS and Linux)
  * SwiftUI Sample (tested on iOS)

## Why LibGodot?

LibGodot has a number of different use cases.
* Control Godot Engine from a host application
  * E.g. Start Godot from a .NET process to use standard .NET tooling
* Use it for automation of development tasks (e.g. building an asset pipeline using a Python GDExtension API) 
* Embed Godot into another application while displaying Godot Windows as part of the host application's UI.

## How to try it out? 

We provide a Github repository with preconfigured build environment, sample apps and more information: https://github.com/migeran/libgodot_project

## Migeran LibGodot Design

The core idea of the LibGodot feature is to build upon Godot's strong extension capabilities: its modularity, powerful type system and the GDExtension API.

The main advantage of this approach is, that this way LibGodot may be used from any language that has a GDExtension binding (e.g. C++, Swift, Rust, Python, ... etc.), with only minimal changes to the binding.

### Godot Instance Lifecycle Management

The main class added by the LibGodot design is the GodotInstance:

```cpp
class GodotInstance : public Object {
	GDCLASS(GodotInstance, Object);

	static void _bind_methods();

	bool started = false;

public:
	GodotInstance();
	~GodotInstance();

	bool start();
	bool is_started();
	bool iteration();
	void shutdown();
};
```

This class is made accessible over the GDExtension API. This class can be used to control a Godot instance.

To actually create a Godot instance a new symbol is added to the GDExtension API:

```cpp
GDExtensionObjectPtr gdextension_create_godot_instance(int p_argc, char *p_argv[], GDExtensionInitializationFunction p_init_func);
```

This function can be used to create a new Godot instance and return a GodotInstance object reference to control it. Both samples show how easy it is to bind this function and then use the generated GDExtension API bindings with the returned GodotInstance object.

To properly destroy a Godot instance the GDExtension API is extended by another symbol:

```cpp
typedef void (*GDExtensionInterfaceDestroyGodotInstance)(GDExtensionObjectPtr p_godot_instance);
```

This function is made available through the GDExtension API's getProcAddress mechanism.

Note: Due to Godot's internal architecture (the use of global data structures and singletons) only one Godot instance may be created in a process.

### Embedding Godot UI

Note: UI embedding is currently implemented for Apple platforms. Please read the [Next Steps](#next-steps) section for more information on the status of other platforms. 

To allow for embedding the Godot UI into a host process the host process needs to be able to pass the necessary data about a native surface where 
Godot may render its contents. The form of this native surface is entirely platform and rendering backend specific.

The Migeran LibGodot design adds the following types to Godot to allow this:

* A new ``DisplayServerEmbedded`` implementation which uses externally provided native surfaces as its rendering targets.
* A ``RenderingNativeSurface`` class and its associated platform specific subclasses, e.g. ``RenderingNativeSurfaceApple``.
* The ``Window`` class is extended by a ``set_native_surface(Ref<RenderingNativeSurface>)`` method which allows specifying the rendering target of a Window in a typesafe, platform independent manner. It is even possible to change the rendering target of a Window dynamically during its lifecycle. 

These classes are all exposed to the GDExtension API, so it is easy to use them from the host process.

The ``DisplayServerEmbedded`` class also exposes methods that allow the injection of input events from the host process into the Godot instance.

The RenderingNativeSurface class hierarchy has an additional function: provides a mechanism using the **Factory Method** design pattern to create compatible RenderingContextDrivers for a native surface.

This allowed us to make the ``DisplayServerEmbedded`` class platform agnostic.

We also refactored the other ``DisplayServer`` implementations to use ``RenderingNativeSurface`` during initialization.

Since all these classes are exposed over the GDExtension API, these can be seamlessly used from any supported language with a GDExtension binding.

### Rationale for RenderingNativeSurface Design

For those who are familiar with Godot Engine internals: there was already a way to pass in platform specific native surface data (called ``WindowPlatformData``) during initialization.

Why is this refactoring necessary to a full fledged reference counted object?

* We chose reference counting because it makes it easier to use on the client side, no need to manage the lifecycle manually. Earlier versions of this PR used simple Godot Objects, but they required manual memory management which was error prone.

* While on Apple platforms it is very easy to pass in a ``CAMetalLayer`` reference, and Metal (and thus MoltenVk) will happily render into it, other platforms impose way more restrictions.
  * For example: On Windows and Linux a separate Vulkan instance and the use of external textures is required to render Godot on a separate thread from the host application's main renderer thread.
  
  * This is not just a theoretical option: We already implemented a prototype on Linux and Windows based on Godot 4.2, where the host process and Godot are using Vulkan External Textures to pass the rendered frames from Godot to the host process. We intend to upgrade and refactor this patch to the current version of the LibGodot patch.
  
  * To use External Textures a thread-safe queue has to be implemented between between the host process and Godot, and a reference-counted RenderingNativeSurface subclass would be an ideal place to implement it, e.g. ``RenderingNativeSurfaceVulkanExternalTextureWin32``.

## Next Steps

### Open for Discussion

We are happy to discuss any part of the design, naming conventions, additional features ... etc.

### Merging for Godot 4.3 as an Experimental Feature

* We propose to merge the current LibGodot patch as an experimental feature in Godot 4.3, because it is already very usable in its current form for many use cases.

### Godot 4.4 Developments

During the Godot 4.4 development cycle additional features may be added, for Example:

* Support for additional platforms for UI embedding: Android, Windows, Linux (both X11 and Wayland).
* Support for OpenGL (Native / ANGLE) UI embedding on all platforms.
* Add a Configuration API that can be used on startup instead of supplying command line parameters. 

## Sponsors & Acknowledgements

* Initial development sponsored by [Smirk Software](https://www.smirk.gg/)
* Rebasing to Godot 4.3 and further development sponsored by [Xibbon Inc.](https://xibbon.com)

* The GDExtension registration of the host process & build system changes were based on @Faolan-Rad's LibGodot PR: https://github.com/godotengine/godot/pull/72883

## About the Authors

This feature was implemented by [Migeran - Godot Engine & Robotics Experts](https://migeran.com).

- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/6267.*

fire commented:
You can fix some of the github ci actions tests with `pre-commit run -a`.

Here are the rest of the errors. https://github.com/V-Sekai/godot/actions/runs/8642913827

dsnopek commented on core/extension/gdextension_manager.cpp:
```diff
@@ -236,6 +236,18 @@ void GDExtensionManager::load_extensions() {
+	if (p_init_func) {
+		Ref<GDExtension> libgodot;
+		libgodot.instantiate();
+		Error err = libgodot->initialize_extension_function(p_init_func, "lib_godot");
+		if (err != OK) {
+			ERR_PRINT("Error initializing extension function");
+		} else {
+			libgodot->set_path("res://__LibGodot");
+			GDExtensionManager::get_singleton()->load_extension("res://__LibGodot");
```
This PR introduces a bunch of new names, and, while I don't necessarily want to start bikeshedding, we probably should discuss if these are the names we want to use, and, in particular, are they internally consistent, and consistent with names already used in other parts of Godot.

So, here we have:

- `"lib_godot"` as the fake entry point name
- `"res://__LibGodot"` as the fake library path. (Also, this needs to be unique among all extensions - with this PR, are developers able to register only one faux GDExtension from their application or multiple? In @Faolan-Rad's PR, if I'm remembering correctly, more than one faux GDExtension could be registered. If multiple are allowed, we'll need a unique path for each.)

And elsewhere we have:

- The `LIBRARY_ENABLED` define
- The "Embedded" feature tag
- The general term "embedded" to describe a faux GDExtension registered through libgodot

These seem a little bit inconsistent, using the terms "libgodot", "embedded" and "library" in different places.

kisg replied:
I agree, and we are open to naming suggestions. I agree that it is important to have consistent naming. Should we just try to use LIBGODOT everywhere as the key for this? Or should we move to the more generic "LIBRARY" name?

dsnopek replied:
Coming back to this because it seems there's still some mixed names (although, it's _much_ improved from the last time I went through this PR).

Personally, I don't like the super generic "library".

I'd be for either "LibGodot" or "embedded Godot", since they are more descriptive, with maybe a slight preference for "LibGodot" because that sounds more like a product/feature name, making it easier to refer to.

If we did go for "LibGodot", then we'd end up with a `LIBGODOT_ENABLED` define, `libgodot://` path, `libgodot` feature tag, and "LibGodot" in the docs and other human-readable places.

I'm surprised there hasn't been more bike-shedding about the name :-)

dsnopek commented on core/extension/gdextension.cpp:
```diff
@@ -787,9 +787,14 @@ Error GDExtension::open_library(const String &p_path, const String &p_entry_symb
+	if (initialization_function(&gdextension_get_proc_address, this, &initialization)) {
 		level_initialized = -1;
 		return OK;
 	} else {
```
GitHub won't let me comment on it directly, because it's not in the diff, but in this `else` block there is this:

```
		ERR_PRINT("GDExtension initialization function '" + p_entry_symbol + "' returned an error.");
		OS::get_singleton()->close_dynamic_library(library);
		return FAILED;
```

We probably don't want to call `OS::get_singleton()->close_dynamic_library(library);` here if `library` is null.

kisg replied:
Good catch, will take a look.

kisg replied:
Moved to GDExtensionLoader implementation.

dsnopek commented on core/extension/gdextension_interface.cpp:
```diff
@@ -1484,6 +1485,20 @@ static void *gdextension_classdb_get_class_tag(GDExtensionConstStringNamePtr p_c
 
+GDExtensionObjectPtr gdextension_create_godot_instance(int p_argc, char *p_argv[], GDExtensionInitializationFunction p_init_func) {
+#ifdef LIBRARY_ENABLED
+	return create_godot_instance(p_argc, p_argv, p_init_func);
```
I think we probably want to prevent creating more than one of these? Godot uses so many singletons, that I suspect we can't make more than one in the same memory space. (Unless this is somehow supported?)

So, this probably should hang on to the instance somewhere, and error here if we attempt to create another instance? And if we're hanging on to it, it'd probably be good to have another interface function to retrieve it.

kisg replied:
Yes, this was omitted by accident, we will add it soon.

kisg replied:
Done.

dsnopek commented on core/extension/gdextension_manager.cpp:
```diff
@@ -236,6 +236,18 @@ void GDExtensionManager::load_extensions() {
 
 	OS::get_singleton()->load_platform_gdextensions();
+
+	if (p_init_func) {
+		Ref<GDExtension> libgodot;
+		libgodot.instantiate();
+		Error err = libgodot->initialize_extension_function(p_init_func, "lib_godot");
```
Rather than passing around the `GDExtensionInitializationFunction p_init_func`, and finally getting around to calling it here, could we simply register the `GDExtensionInitializationFunction` into some list (a list assuming we could have multiple) and then loop over them here? (Again, this depends on if we want to allow multiple or not.)

fire replied:
I would prefer to be able to use multiple godot gdextension symbols in the same libgodot binary as a user.

kisg replied:
As I mentioned above, we had support for it in an earlier version of the patch, we will resurrect it in the next version of this PR.

kisg replied:
Added a new load_function_extension() method exposed to GDExtension in the GDExtensionManager that is using the new function loader internally.

dsnopek reviewed (commented):
Thanks! I'm so excited to finally see this PR :-)

I looked only at the GDExtension-related changes so far. There's a bunch of rendering and display server related changes that I haven't looked at.

At high-level, this seems like a pretty good approach. Also, I really appreciate that the code changes to GDExtension are fairly minimal. :-)

On PR https://github.com/godotengine/godot/pull/72883, we discussed the possibility of introducing a new concept, perhaps called `GDExtensionLoader`, that would encapsulate the loading behavior, so that we could have `GDExtensionLibraryLoader` that used `OS::open_dynamic_library()` (so the current way) and then maybe a `GDExtensionEmbeddedLoader` that would just take an initialization function.

However, given how simple the changes are here, I think we could perhaps save something like that for a future refactor? It could make things a little cleaner by not having to constantly check if `GDExtension::library` is null or not, though.

fire commented:
I've tried to get the Godot Engine Github actions continuous integration to pass. Most tests pass except for Windows. https://github.com/V-Sekai/godot/tree/vsk-libgodot-migeran-4.3

dsnopek commented on core/register_core_types.cpp:
```diff
@@ -266,6 +267,7 @@ void register_core_types() {
 	GDREGISTER_ABSTRACT_CLASS(ResourceImporter);
 
 	GDREGISTER_CLASS(GDExtension);
+	GDREGISTER_CLASS(GodotInstance);
```
Actually, thinking about this a little more: Do we even need to have a bound `GodotInstance` class?

There should only every be 1 instance of it. We only want to ever use it from GDExtension (not scripts). It has only 4 methods that we need to call.

Couldn't we instead just add 4 functions to `gdextension_interface.h`?

Maybe we'll still want a class internally, but I don't think we necessarily need to bind a class for this. Not binding it avoids the whole issue with controlling its construction and hiding it from script.

kisg replied:
This is a bug, we wanted to declare it as GDREGISTER_ABSTRACT_CLASS() like we did with the DisplayServerEmbedded class. It effectively disables the creation function in the API, so it will not be possible to create an object except with the create_instance function.

There are a couple reasons why I would like to keep it as a GDExtension class:
- I see this as a first (tiny) step to turn Godot into a library where you can create multiple instances. Right now, it will be limited to one instance (as you already remarked below), but I see no reason why we shouldn't have an API that could allow it in the future without any changes on the client side.
- Registering it as GDExtension will reduce the burden on the language binding developers: no need to bind it manually, the generator will take care of it. We already demonstrate this for Swift and C++ in the samples.
- My long(er) term idea is that we should change the way Godot is initialized by extending GodotInstance:
  - On create_instance() we would only run a stripped down version of Main::setup() without the command line parsing and configuration.
  - GodotInstance then could provide a Configuration API, to configure the instance before start() is called. Of course we can provide a configure(String[] args) method that will call the command line parser.
  - Then when start() is called, the rest of the startup sequence would be executed.
  - GodotInstance could also add the APIs to register/unregister embedded GDExtensions. In this case we would not have to add separate GDExtension function pointers for these, which would have to be handled by the language binding developers manually.
  - My hope is that this extended API of the GodotInstance class would be used for configuring, starting and controlling the Godot instance on all platforms - either through the GDExtension API or linking directly to it inside the engine. This could also reduce the fragmentation (and the required code) across the platform implementations.

dsnopek replied:
Ok, thanks for the explanation!

I'm not totally opposed to registering a `GodotInstance` class. But when we have something we don't want available to scripting and/or want to tightly control, we frequently avoid registering/exposing it, and add functions to the GDExtension interface instead, and this kind of feels like one of those cases.

If we do register the `GodotInstance` class, we just need to make sure it can't be created or used in ways we don't want, and, _especially_ that scripting can't access it.

kisg replied:
Now GodotInstance is registered as abstract and the only place it is instantiated in libgodot_create_godot_instance()

kisg commented:
> Thanks! I'm so excited to finally see this PR :-)
> 
Thank you, and thank you for taking the time to review it!

> I looked only at the GDExtension-related changes so far. There's a bunch of rendering and display server related changes that I haven't looked at.
> 
> At high-level, this seems like a pretty good approach. Also, I really appreciate that the code changes to GDExtension are fairly minimal. :-)
> 
> On PR #72883, we discussed the possibility of introducing a new concept, perhaps called `GDExtensionLoader`, that would encapsulate the loading behavior, so that we could have `GDExtensionLibraryLoader` that used `OS::open_dynamic_library()` (so the current way) and then maybe a `GDExtensionEmbeddedLoader` that would just take an initialization function.
> 
> However, given how simple the changes are here, I think we could perhaps save something like that for a future refactor? It could make things a little cleaner by not having to constantly check if `GDExtension::library` is null or not, though.

I think introducing the concept of library loaders would benefit not just this use case, but also other language support that are implemented as GDExtensions. If you recall, I proposed a feature like that in my .NET integration proposal:  https://docs.google.com/document/d/1QwZpo3oIKHyita1mDOIB_xIxdwJ5rtAmpY4_vP9S02Y/edit?usp=sharing

However, I am also not sure if we should try to push this whole loader concept into this PR, we could do that in a follow-up PR. We can discuss this more in the GDExtension meeting.

kisg commented on core/extension/gdextension_interface.h:
```diff
@@ -2864,6 +2864,11 @@ typedef void (*GDExtensionsInterfaceEditorHelpLoadXmlFromUtf8Chars)(const char *
  */
 typedef void (*GDExtensionsInterfaceEditorHelpLoadXmlFromUtf8CharsAndLen)(const char *p_data, GDExtensionInt p_size);
 
+typedef GDExtensionObjectPtr (*GDExtensionInterfaceCreateGodotInstance)(int p_argc, char *p_argv[], GDExtensionInitializationFunction p_init_func);
+typedef void (*GDExtensionInterfaceDestroyGodotInstance)(GDExtensionObjectPtr p_godot_instance);
```
We have the GDExtension initialization callback in the create_instance function as a parameter so the language bindings in the host application can be initialized before the function returns. You can check out how we do this in the C++ sample app: https://github.com/migeran/libgodot_project/blob/26fc59a757d4b31691bbf15e88c909cc921158b3/samples/cpp_sample/src/main.cpp#L74
(Note: that if you agree, we would be happy to move this LibGodot class (or a variant of it) to godot-cpp for an even better developer experience.)

We can definitely add separate load / unload embedded extension functions. (We actually had a variation of these in a previous unreleased incarnation of this patch, we just wanted to keep the PR as light as possible. We are happy to add these back.)

kisg replied:
These have been resolved by the latest update that was built on top of GDExtensionLoader.

kisg commented on core/extension/gdextension_interface.h:
```diff
@@ -2864,6 +2864,11 @@ typedef void (*GDExtensionsInterfaceEditorHelpLoadXmlFromUtf8Chars)(const char *
+typedef GDExtensionObjectPtr (*GDExtensionInterfaceCreateGodotInstance)(int p_argc, char *p_argv[], GDExtensionInitializationFunction p_init_func);
+typedef void (*GDExtensionInterfaceDestroyGodotInstance)(GDExtensionObjectPtr p_godot_instance);
+
+GDExtensionObjectPtr gdextension_create_godot_instance(int p_argc, char *p_argv[], GDExtensionInitializationFunction p_init_func);
```
Omitting this function declaration works, if we compile Godot as a shared library, because we can look up the symbol by name with dlsym().

However, we also support building LibGodot as a static library, (we use it now on iOS for example, but there is nothing that would prevent using it on other platforms, if desired by the developer). In such case they need the function declaration.

We contemplated adding a separate libgodot.h header for this, but we felt it would be an overkill to add a new header file for a single function declaration.

We could move this declaration into a #ifdef LIBGODOT_STATIC block to make sure that we don't add the declaration if it is not strictly necessary. What do you think?

dsnopek replied:
The function declaration would definitely need to be in a separate header - wrapping the declaration in a macro isn't sufficient.

We don't treat `gdextension_interface.h` as "real C code" because not all bindings will be including it with a C compiler - instead they may generate code from it using a custom utility, so we have a number of restrictions on what can be in that file.

(As an aside: this relates to a [recent proposal](https://github.com/godotengine/godot-proposals/discussions/9524) to instead generate `gdextension_interface.h` from an XML or JSON file. This hasn't been discussed by the GDExtension team yet and no one is working on it as far as I know, but it would be a nice way to formalize "`gdextension_interface.h` isn't real C code" and prevent contributors from trying to add stuff there that isn't allowed :-))

kisg replied:
These have been resolved by the latest update that was built on top of GDExtensionLoader.
The gdextension_interface.h is not touched at all, and we introduced a new libgodot.h that introduces the create and destroy instance function declarations.

kisg commented:
> I've tried to get the Godot Engine Github actions continuous integration to pass. Most tests pass except for Windows. https://github.com/V-Sekai/godot/tree/vsk-libgodot-migeran-4.3
> 
> Edited: Time is limited for me to work on this, so I hope this can help your efforts.

Thank you @fire ! Would it be ok if I squash your changes into a single commit, and add it to our branch?

fire commented:
Do what what works for you.

GeorgeS2019 commented:
@fire 

Are you familiar with the DirectX surface so the UI is not just a desktop BUt standard userControl  compatible with e.g. Avalonial or WPF or WinForm?
https://github.com/migeran/libgodot_project/commit/246cb91cd8a4255f605653869a912d6076a5efa7

fire commented:
For the purposes of this libgodot pull request I am postponing these for the future as listed in the start:

    * WPF (w/ WPF DirectX 12 Extensions)
    * Support for OpenGL (Native / ANGLE) UI embedding on all platforms.
    * Add a Configuration API that can be used on startup instead of supplying command line parameters.

There are several unfinished review comments from @dsnopek to be worked on.

AThousandShips commented on platform/linuxbsd/godot_runtime_api.cpp:
```diff
@@ -0,0 +1,69 @@
+
+class GodotInstanceLinuxBSD : public GodotInstance {
+public:
+	bool start() override {
```
```suggestion
	virtual bool start() override {
```

AThousandShips commented on platform/linuxbsd/godot_runtime_api.cpp:
```diff
@@ -0,0 +1,69 @@
+		return result;
+	}
+
+	void shutdown() override {
```
```suggestion
	virtual void shutdown() override {
```

AThousandShips commented on platform/linuxbsd/x11/rendering_context_driver_vulkan_x11.cpp:
```diff
@@ -31,6 +31,8 @@
 #ifdef VULKAN_ENABLED
 
 #include "rendering_context_driver_vulkan_x11.h"
+#include "drivers/vulkan/rendering_native_surface_vulkan.h"
```
```suggestion

#include "drivers/vulkan/rendering_native_surface_vulkan.h"
```

AThousandShips commented on platform/macos/godot_runtime_api.mm:
```diff
@@ -0,0 +1,72 @@
+
+class GodotInstanceMacOS : public GodotInstance {
+public:
+	bool start() override {
```
```suggestion
	virtual bool start() override {
```

AThousandShips commented on platform/macos/godot_runtime_api.mm:
```diff
@@ -0,0 +1,72 @@
+		return result;
+	}
+
+	void shutdown() override {
```
```suggestion
	virtual void shutdown() override {
```

kisg commented:
Thank you all for your reviews and contributions. We are working on some updates and will include the contributed features as well. Stay tuned!

octanejohn commented:
does this help with running/embedding the game in editor
related
https://github.com/godotengine/godot-proposals/issues/7213
https://github.com/godotengine/godot-proposals/issues/9142

dsnopek commented:
@octanejohn:

> does this help with running/embedding the game in editor

I don't think so. This would allow other applications to embed Godot, but I don't think it would allow Godot to embed Godot. Those proposals offer different approaches to accomplish their goals

fpdotmonkey commented:
> This would allow other applications to embed Godot, but I don't think it would allow Godot to embed Godot.

Genuine question, what's the difference?

kisg commented:
With this latest update most review comments should be fixed.

Still on the TODO list:
- Windows build
- Fix CI builds
- Create a .NET/C# example using the new GDExtension-based .NET implementation.

GeorgeS2019 commented:
> Create a .NET/C# example using the new GDExtension-based .NET implementation.

See https://github.com/godotengine/godot/pull/72883#issuecomment-2226296833

YakoYakoYokuYoku reviewed (changes requested):
Overall, this is a pretty good job, there's some suggestions I've made to improve this PR.

AThousandShips commented:
@YakoYakoYokuYoku Where did you leave suggestions? I can't see any code suggestions

GeorgeS2019 commented:
@AThousandShips 

The [**windows build** PR](https://github.com/migeran/libgodot_project/pull/6) is ready for merge
I think this PR is ready to merge too. Please feedback

akien-mga commented:
@GeorgeS2019 We're in feature freeze for 4.3, and this PR is slated for 4.4. Please be patient.

YakoYakoYokuYoku commented on scene/main/window.cpp:
```diff
@@ -835,6 +853,11 @@ void Window::hide() {
 void Window::set_visible(bool p_visible) {
 	ERR_MAIN_THREAD_GUARD;
+	if (native_surface.is_valid()) {
+		visible = true;
```
I understand that the native surface is handled somewhere else, but there has to be a way to tell if its window is hidden or not. The `is_visible` method can be modified to check the native surface (here or in another PR) and this assignment to `true` can be dropped.

YakoYakoYokuYoku commented on servers/display_server_embedded.cpp:
```diff
@@ -0,0 +1,540 @@
+			ERR_PRINT(vformat("Failed to initialize %s context", rendering_driver));
+			memdelete(rendering_context);
+			rendering_context = nullptr;
+			return;
```
Can `r_error` be set here?

YakoYakoYokuYoku commented on servers/display_server_embedded.cpp:
```diff
@@ -0,0 +1,540 @@
+
+#if defined(GLES3_ENABLED)
+	if (rendering_driver == "opengl3") {
+		ERR_FAIL_MSG("Embedded OpenGL rendering is not yet supported.");
```
Before this line `r_error` could be set to `ERR_UNAVAILABLE` or another value.

YakoYakoYokuYoku commented on servers/display_server_embedded.cpp:
```diff
@@ -0,0 +1,540 @@
+#if defined(VULKAN_ENABLED)
+	drivers.push_back("vulkan");
+#endif
+	//#if defined(GLES3_ENABLED)
+	//	drivers.push_back("opengl3");
+	//#endif
```
Maybe add a `TODO` comment.

YakoYakoYokuYoku commented on servers/display_server_embedded.cpp:
```diff
@@ -0,0 +1,540 @@
+	if (p_id != INVALID_WINDOW_ID) {
+		_window_callback(input_event_callbacks[p_id], p_event);
+	} else {
+		for (const KeyValue<WindowID, Callable> &E : input_event_callbacks) {
```
This `E` could be changed for something more meaningful, like `cb` or something.

fire replied:
E is the typical coding style of godot engine, but can be better!

YakoYakoYokuYoku commented on servers/display_server_embedded.cpp:
Some characteristics, like DPI, maximum and minimum size, focus, plus many more can be queried from the native surface. For now `TODO` comments can suffice as what I've said gets out of the scope of this PR.

YakoYakoYokuYoku commented:
> @YakoYakoYokuYoku Where did you leave suggestions? I can't see any code suggestions

I've messed up with the PR review, my bad. Regardless, here they go.

rovantiq commented:
> this PR is slated for 4.4

@akien-mga Is the plan to ship 4.4 with the unified editor (assuming things go well with this and related PRs)?

fire commented on core/extension/gdextension_library_loader.cpp:
```diff
@@ -0,0 +1,399 @@
+		// Iterate the libraries, finding the best matching tags.
+		String best_library_path;
+		Vector<String> best_library_tags;
+		for (const String &E : libraries) {
```
E is the standard godot C++ style.

fire commented on core/extension/godot_instance.h:
Do you know how to force git do this operation? This is a automatic action.

kisg commented:
> > this PR is slated for 4.4
> 
> @akien-mga Is the plan to ship 4.4 with the unified editor (assuming things go well with this and related PRs)?

@rovantiq Could you please point me to the unified editor PRs / proposals that you are referring to?

Gnumaru commented:
> @rovantiq Could you please point me to the unified editor PRs / proposals that you are referring to?

Maybe rovantiq was refering to the possibility of related PRs, not to specific PRs that he could remember. I remember myself this topic of having an unified .net editor but I can't remember where I first read about it. I have just searched and found a mention about a .net contributors meeting here https://github.com/godotengine/godot-proposals/issues/7895#issuecomment-1737839064, I'm just copy-pasting the text here for convenience.

==========

> Since I don't see it mentioned anywhere, I just want to paste below the minutes from last week's .NET contributors meetings regarding this specific topic. Note that I'm just quoting, I don't personally share the same opinion on all of the points.
> 
>> Editor unification (.NET integration as a plugin/extension) (switch to integration as gdextension or just always bundle it)
>> 
>> - Good way to move forward. Decouple C# support from the core. If anything is missing in GDExtension, it would be missing for other languages too.
>> - Would need to drop the entire interop code and rewrite it from scratch. Module uses a number of APIs which might not be exposed to GDExtension yet.
>> - C++ extension would not needed to be reloaded, and it would be able to reload C# assemblies just like the module does currently.
>> - Performance difference need to be assessed - attendees not convinced this would make a difference to the current approach.

akien-mga commented:
The unified editor with .NET support via GDExtension is being worked on (see https://github.com/raulsntos/godot-dotnet), but this is orthogonal to this proposal and thus off-topic.

GeorgeS2019 commented:
https://github.com/godotengine/godot/pull/90510#issuecomment-2159522891
```
Still on the TODO list:

- Fix CI builds
```
##### [WIP 18 Aug 2024] There is now working example
https://github.com/migeran/libgodot_project/issues/1#issuecomment-2294996680

- Windows build
- Create a .NET/C# example using the new GDExtension-based .NET implementation.

dsnopek commented on core/extension/gdextension_manager.h:
```diff
@@ -63,6 +66,7 @@ class GDExtensionManager : public Object {
 
 public:
 	LoadStatus load_extension(const String &p_path);
+	LoadStatus load_function_extension(const String &p_path, GDExtensionConstPtr<const GDExtensionInitializationFunction> p_init_func);
```
Another naming nitpick: "function extension" is an unclear term that we don't really use anywhere else.

Maybe `load_extension_from_function()`?

Or, if we can decide on an overall name for this feature we could use that term. So, if we standardize on "LibGodot", then this could be `load_libgodot_extension()`? Or, if we standardize on "embedded", then it could be `load_embedded_extension()`?

dsnopek commented on platform/android/api/api.cpp:
```diff
@@ -34,11 +34,21 @@
 
+void register_core_android_api() {
+#if defined(ANDROID_ENABLED)
+	GDREGISTER_ABSTRACT_CLASS(RenderingNativeSurfaceAndroid);
```
Will a stub class for `RenderingNativeSurfaceAndroid` be registered when _not_ on Android? Ideally, we'd want a class the same name and all the same methods available on all platforms, but only a version that actually works when on Android.

We do a similar thing for `JNISingleton`, `JavaClassWrapper`, and `JavaClass` - those are always registered (although, sometimes in a different way) on all platforms, but when not on Android, their real functionality is removed.

I haven't looked at the other platforms, but we'd want to do the same for them as well. Or, alternatively, we could just expose the base class and a factory method, and then developers never need to interact with the classes for the individual platforms, just the base class (which is sort of like what we do for `DisplayServer`).

The point is just that we don't want classes to appear/disappear when on different platforms. Otherwise, developers get errors depending on which platform they opened their project up on, just because they refer to the class name in their GDScript.

dsnopek reviewed (commented):
From the perspective of GDExtension, this PR is looking really good now!

Now that it's built on top of the Raul's "GDExtension loaders" PR, the changes here are really pretty minimal.

I'm still not crazy about exposing `GodotInstance`, but now that it's abstract, there shouldn't be any risk of developers getting ahold of one when they shouldn't. I think I can live with that.

At this point, I think it's primarily the rendering and platform-related code that needs review and discussion.

AndrewBabbitt97 commented:
Also curious about if it will be published on NuGet.

My use case involves deserializing Godot resources from an ASP.NET Core project so we can read values easily from tres/res files. Having this in a NuGet package would make consumption a breeze.

GeorgeS2019 commented:
@mororo250

Here is where to push.
https://github.com/migeran/libgodot_project/tree/windows_build
![image](https://github.com/user-attachments/assets/c2a19d60-9ad7-4361-a23e-fac40bd30e68)

kisg commented:
Actually we already have this updated for current master, just needs some cleanup and will go out "Real Soon Now".

Gnumaru commented:
"Real Soon Now" should probably be "after" 4.4 stable is released, right?

GeorgeS2019 commented:
@Gnumaru
 @kisg
 
> "Real Soon Now" should probably be "after" 4.4 stable is released, right?

==> It means indefinite, and definitely not soon.

![image](https://github.com/user-attachments/assets/274a61d2-ccdf-4eea-921f-4c92a86d573b)

AThousandShips commented:
It doesn't meed indefinate, it means "no date set", because this still need work to be fixed, and hasn't been approved yet, it will be put on what ever milestone is the current one when ready to merge!

tbibre commented:
Any hope to see this as an experimental feature in 4.5 ?😊

dsnopek commented:
> Any hope to see this as an experimental feature in 4.5 ?😊

I would really love for this to be merged for Godot 4.5! And right now is actually a great time in the dev cycle to do it.

However, this PR still needs some work:

- It's currently failing CI
- I left [some review](https://github.com/godotengine/godot/pull/90510#pullrequestreview-2298856292) that still hasn't been addressed

Also, I think the rendering team still has to do a final sign off on the rendering changes, however, they may be waiting on the two previous points being addressed first

In any case, it's up to @kisg to finish up or hand off to someone else to do so

adamscott commented:
> However, this PR still needs some work:
> 
>     * It's currently failing CI
> 
>     * I left [some review](https://github.com/godotengine/godot/pull/90510#pullrequestreview-2298856292) that still hasn't been addressed

The last commit dates back to Jul 3, 2024. Some work is needed to rebase master onto this branch.

kisg commented:
Sorry for the delay in updating this PR. I just pushed a version updated to 4.4.1, you can check the current state there: https://github.com/migeran/godot/tree/libgodot_44_stable

We also have in progress versions of Windows / Linux embedding + Android that are not in that PR yet, but will be also released later.

GeorgeS2019 commented:
Window version would be really appreciated

zorbathut commented:
@kisg, I have some immediate need for this in my own project. Would it be helpful if I fixed up the various CI failures? I'd personally like to get this upstreamed and am willing to invest a bit of elbow grease in the process.

Am I correct in thinking that `libgodot_44_stable` is the right one to work off, or should I be working off some other branch?

Should I send you a pull request? :V

fire commented:
As far as I know the biggest blocker to merging is that the cicd tests don't work for all platforms.

kisg commented:
@zorbathut 

> @kisg, I have some immediate need for this in my own project. Would it be helpful if I fixed up the various CI failures? I'd personally like to get this upstreamed and am willing to invest a bit of elbow grease in the process.
> 
> Am I correct in thinking that `libgodot_44_stable` is the right one to work off, or should I be working off some other branch?
> 
> Should I send you a pull request? :V

This would be great thank you. Yes, please work against libgodot_44_stable for now.

zorbathut commented:
Pull request: https://github.com/migeran/godot/pull/3

Github test report: https://github.com/zorbathut/godot/actions/runs/15233766019

There's a bunch of notes in the pull request regarding the changes themselves.

One thing of note is that the build still fails on iOS and macOS. I believe this is bitrot regarding the testing framework; iOS has a build failure that I suspect is caused by a newer compiler, macOS straight-up doesn't even start building. I suspect these have been fixed in Godot 4.5, but this branch is explicitly a 4.4 branch, and I don't want to start cherrypicking in 4.5 stuff because that's just going to make any ultimate master-branch merge that much more annoying.

I think if the pull request is accepted, the next step is to make a `libgodot_master` branch and rebase the entire shebang on top of `master`, so it will actually merge in cleanly and isn't subject to months-old github-actions bitrotting. I am also happy to do this, once the 4.4 branch is in reasonably good shape.

I did some testing but not massive testing; I can vouch for some of the Godot demo projects working fine under Linux. I don't currently have a good way to test the actual libgodot functionality.

fire commented:
The standard procedure is for pull requests is to target master. Then if possible a backport is done.

fire commented:
Some thoughts on these efforts

Solidifying it for kisg's `4.4-stable-targeting` branch first before moving to `master` definitely sounds like the smoother path. Good luck with it!

**Targeting 4.5 Godot Engine feels ambitious, given the early June feature freeze (very soon), especially with the outstanding rendering review and the reported rendering-related issues. It would benefit diagnosis if anyone can provide an MRP (Minimum Reproduction Project) for those rendering issues.**

That said, I'm optimistic that with dedicated effort – perhaps including some focused 'salvage' work on the existing work by interested contributors – it can reach a mergeable state (all GitHub Actions passing) for 4.6+. I'm happy to be part of that contributing effort.

fire commented:
@Zorbathut, It's Tuesday, and I'm ready to pick this PR back up. Do you have any updates on the new branch, or are there specific areas you'd like me to focus on or clarify?

kisg commented:
I am in the middle of refactoring this for 4.5 + reviewing https://github.com/migeran/godot/pull/3.

GeorgeS2019 commented:
> The other downside is that getting this all building and working is nontrivial. I can't just say "go do the thing", I need to provide an entire framework for making it work, similar to the (no longer up-to-date) libgodot_llgo_demo repo by @JiepengTan.

@zorbathut 

Thanks for realising it...it is the only quick hack that demonstrates it is feasible for windows and c#

GeorgeS2019 commented:
>  I may be wrong on this, but I think they were trying to make Godot runnable from inside Unreal Engine's C++ environment?

Very strange ..hearing it first time...what is the use case?

zorbathut commented:
> Thanks for realising it...it is the only quick hack that demonstrates it is feasible for windows and c#

For what it's worth, if you run into trouble, please let me know; it's been tried on exactly three computers and two of them are mine, so it can definitely use more testing. But it *should* be extremely easy to get working.

(. . . okay, "extremely easy" relative to the kind of person who's likely to be browsing godot github pull requests)

> Very strange ..hearing it first time...what is the use case?

Judging from [their blog page](https://migeran.com/blog/ultimate-asset-pipeline-for-godot-engine) it sounds like they're trying to set up an asset import pipeline that can interact simultaneously with both Unreal/Unity/Blender and Godot code, which is honestly a sorta cool idea and avoids having to use an intermediate format.

GeorgeS2019 commented:
@zorbathut 

The .net community face a few challenges..moving along this path

It would be great you share your envisioned coming challenges for different platform, e.g. mobile, web, AOT, wasm...

Have you thinking ahead. Clearing mind field ..

What is your view, if we have choice, with wasm... currently blocked by .dotnet team ...why we use mono approach for web export..
For Dotnet

I do not have the overview...appreciate if you could share your vision.
.thx for your passion... appreciate

zorbathut commented:
> I do not have the overview...appreciate if you could share your vision.

I don't have one. Sorry!

I'm vaguely interested in webasm C# working, which is why I'm keeping an eye on it, but it's not a high priority for me. Perhaps if this gets upstreamed I'll take a crack at it eventually.

I am not part of the Godot team and have no insight on their decisions.

GeorgeS2019 commented:
@zorbathut 

What the .net Godot community needs is someone like you, helping us to explore ahead and share with us what could be questions ahead.

I do not get the issue you raised in terms of efficient unit testing of libgodot.dll, bring able to start and stop an instance ??? 

Thx for raising your observation, the community will discuss if your point is relevant and valid.

kisg commented:
> > Thanks for realising it...it is the only quick hack that demonstrates it is feasible for windows and c#
> 
> For what it's worth, if you run into trouble, please let me know; it's been tried on exactly three computers and two of them are mine, so it can definitely use more testing. But it _should_ be extremely easy to get working.
> 
> (. . . okay, "extremely easy" relative to the kind of person who's likely to be browsing godot github pull requests)
> 
> > Very strange ..hearing it first time...what is the use case?
> 
> Judging from [their blog page](https://migeran.com/blog/ultimate-asset-pipeline-for-godot-engine) it sounds like they're trying to set up an asset import pipeline that can interact simultaneously with both Unreal/Unity/Blender and Godot code, which is honestly a sorta cool idea and avoids having to use an intermediate format.

That was a first demo years ago of a very early LibGodot implementation, and definitely not the "endgame", or main use case. 

Since then LibGodot has been used in production in iOS, Android and Windows apps already with support for C++, Java, Swift and not yet disclosed (soon...) languages.

dylanthesoldier17 commented:
> > I do not have the overview...appreciate if you could share your vision.
> 
> I don't have one. Sorry!
> 
> I'm vaguely interested in webasm C# working, which is why I'm keeping an eye on it, but it's not a high priority for me. Perhaps if this gets upstreamed I'll take a crack at it eventually.
> 
> I am not part of the Godot team and have no insight on their decisions.

A side note is if libgodot capability is achieved then it should allow embedding into MAUI projects with the TFMs for Android and iOS. That should allow access to all native libraries Microsoft has built in MAUI. No need to build extensions in swift to access things like CloudKit or Java on Android.

GeorgeS2019 commented:
@dylanthesoldier17 

https://youtu.be/L06KBOWCsSk?si=bsWMvCnOGTGhq_7e

Watch the Displayserver to understand what possibilities available for MAUI, and older e.g. WPF and winform

dylanthesoldier17 commented:
> @dylanthesoldier17
> 
> https://youtu.be/L06KBOWCsSk?si=bsWMvCnOGTGhq_7e
> 
> Watch the Displayserver to understand what possibilities available for MAUI, and older e.g. WPF and winform

Watched the frontend display server section and the display server portion is application specific.

That likely shouldn’t be an issue since the API calls are all done in all .NET code bridged into the native libraries through Microsoft’s shim.

As far as compatibility is concerned I’d assume if they (people in this PR) are working on merging their solution to embed Godot as a library then I believe there wouldn’t be any major loss in functionality and would operate the same. The nuances would be the embedding portion as outlined in video. 

Is that what you’re pointing out?

ArchercatNEO commented:
If you want faster communication/more people official godot discussions happen in https://chat.godotengine.org/home you could ping fire over there after finding an appropriate channel

fire commented:
https://linktr.ee/chibifire has my discord contact info. Please join the discord server since I exceeded the 1000 friend maximum of Discord :(.

Fundamentally the blocker is that it needs to pass all the operating system tests.

To have continual progress we would start with all the platforms disabled. So the tests are starts green.

It needs to work on all the platforms we currently test in CICD.

However, since we already have support for Linux and Macos. I would expect those to pass easily.

Would suggest `Add functional LibGodot OS bindings for Linux` first because even though I'm on Windows many of the development team are on Windows.

Would suggest second `Add functional LibGodot OS bindings for Apple` because libgodot was fully completed because of an ipad port of Godot Engine editor.

Our V-Sekai project wants `Add functional LibGodot OS bindings for Windows` and `Add functional LibGodot OS bindings for Android` because we're PCVR and mobile VR dedicated.

Would expect either we do `Add language-specific support for C#` at the very start because C# build would be part of the very first pass in the github actions.

Chat anytime.

kisg committed:
LibGodot: Fix PCRE2 duplicate symbol for iOS build

kisg committed:
LibGodot: Expose method static and hash through get_method_list

kisg committed:
LibGodot: Implement extended logging

kisg committed:
LibGodot: Shader includes should be handled as proper dependencies for export

kisg committed:
LibGodot: Prerequisites for reinitialization support

- Adds missing singleton destructions
- Extends ClassDB & ObjectDB with reinitialization support

kisg committed:
LibGodot: Fix audio driver reinit

kisg committed:
LibGodot: Support disabling CoreAudio

kisg committed:
LibGodot: Metal Simulator Support

kisg committed:
LibGodot: Main Feature

* Add a new GodotInstance GDCLASS that provides startup and iteration commands to control a Godot instance.
* Adds a libgodot_create_godot_instance entry point that creates a new Godot instance and returns a GodotInstance object.
* Adds a libgodot_destroy_godot_instance entry point that destroys the Godot instance.
  * Now also supports creating a new Godot instance after the previous one was destroyed.
  * Keeps the StringName tables across instances.
* Allow specifying an external native rendering surface handle to render Godot into the UI of a host application.
  * It is also possible to embed multiple Godot windows into the UI of a host application.
  * Currently supported on MacOS and iOS
* Adds support for running Godot on a background thread

Sample Apps: https://github.com/migeran/libgodot_project

Developed by [Migeran](https://migeran.com)

Sponsors & Acknowledgements:

* Initial development sponsored by [Smirk Software](https://www.smirk.gg/)
* Rebasing to Godot 4.3 and further development sponsored by [Xibbon Inc.](https://xibbon.com)
* The GDExtension registration of the host process & build system changes were based
  on @Faolan-Rad's LibGodot PR: https://github.com/godotengine/godot/pull/72883

Co-Authored-By: Gabor Koncz <gabor.koncz@migeran.com>

kisg committed:
LibGodot: Implement external target rendering for OpenGL and Vulkan.

kisg commented:
Just a quick note:

I just pushed an update to our libgodot PR, rebased to the current master, which includes many new features, like:
- start / stop support
- Metal simulator support
- Embedding support using external textures / shared OpenGL contexts (required on platforms like Windows and Linux X11 and for toolkits like Qt/QML)

Additional features like Android support already work as well (as you have seen in our GodotCon presentation), but we intend to submit that as a separate PR, because this PR is becoming pretty huge already. We are also open to submit features not directly related to LibGodot (like Metal Simulator support) separately, but we did not want to delay this update any longer.

We are working on updating the samples + adding new ones in https://github.com/migeran/libgodot_project, you can expect them there next week, will update you here when ready. 

We (the Migeran team) intend to attend the upcoming Core, GDExtensions and Rendering meetings in the coming weeks where we can discuss the details of this PR. 

Let's get LibGodot into 4.6!

clayjohn commented on servers/rendering/rendering_native_surface_external_target.cpp:
```diff
@@ -0,0 +1,442 @@
+/**************************************************************************/
+/*  rendering_native_surface_external_target.cpp                          */
```
Can you provide some context to why this exists?

Right off the bat, its a huge red flag to have so much OpenGL code leaking out of the driver folder. 

I am very wary of having an entire system that appears to duplicate what already exists, but sitting in an entirely different part of the code base. I suspect that this design (whatever its intended purpose) will be an enormous maintenance burden and will basically result in LibGodot being extremely brittle and prone to breaking from seemingly unrelated changes

clayjohn commented on servers/rendering/shader_compiler.cpp:
```diff
@@ -300,6 +303,10 @@ String ShaderCompiler::_get_sampler_name(ShaderLanguage::TextureFilter p_filter,
 	return String(name_mapping[p_filter + (p_repeat == ShaderLanguage::REPEAT_ENABLE ? ShaderLanguage::FILTER_DEFAULT : 0)]);
 }
 
+bool ShaderCompiler::_is_cube_map_array_function(const String &p_function_name) {
+	return p_function_name == "textureSize" || p_function_name == "texture" || p_function_name == "texture" || p_function_name == "textureLod" || p_function_name == "textureGrad";
+}
+
```
This function should be replaced with an LUT

clayjohn commented on servers/display_server.cpp:
```diff
@@ -1602,6 +1643,12 @@ void DisplayServer::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("set_native_icon", "filename"), &DisplayServer::set_native_icon);
 	ClassDB::bind_method(D_METHOD("set_icon", "image"), &DisplayServer::set_icon);
 
+	ClassDB::bind_method(D_METHOD("mouse_button", "x", "y", "mouse_button_index", "pressed", "double_click", "cancelled", "window"), &DisplayServer::mouse_button);
+	ClassDB::bind_method(D_METHOD("mouse_motion", "prev_x", "prev_y", "x", "y", "window"), &DisplayServer::mouse_motion);
+	ClassDB::bind_method(D_METHOD("touch_press", "idx", "x", "y", "pressed", "double_click", "window"), &DisplayServer::touch_press);
+	ClassDB::bind_method(D_METHOD("touch_drag", "idx", "prev_x", "prev_y", "x", "y", "pressure", "tilt", "window"), &DisplayServer::touch_drag);
+	ClassDB::bind_method(D_METHOD("key", "key", "char", "unshifted", "physical", "modifiers", "pressed", "window"), &DisplayServer::key);
+
```
I'm very skeptical about handling input through a DisplayServer. Can you provide some context why this is needed and can't use the usual input handling route?

clayjohn commented on scene/resources/3d/sky_material.cpp:
```diff
@@ -291,6 +291,8 @@ void ProceduralSkyMaterial::cleanup_shader() {
 			RS::get_singleton()->free(shader_cache[i]);
 		}
 	}
+	shader_cache[0] = RID();
```
These seem unrelated to the PR. Does this fix an unreported bug or something?

clayjohn commented on scene/resources/shader.cpp:
```diff
@@ -100,11 +100,23 @@ void Shader::set_code(const String &p_code) {
+		String error_pp;
+		List<ShaderPreprocessor::FilePosition> err_positions;
+		Error result = preprocessor.preprocess(p_code, path, preprocessed_code, &error_pp, &err_positions, nullptr, &new_include_dependencies);
+		if (result != OK) {
```
Can you provide some background on this change. I thought that we already handled error printing from includes in the normal shader error handling code. 

For context, users find it very annoying to get spammed by shader include error thats are handled elsewhere, so I would avoid spamming them more without a really good reason

clayjohn reviewed (commented):
I've just taken a quick looks so far. But I can say for certain that this PR is not ready for serious PR review yet. There are hundreds of files touched with changes unrelated to the actual content of the PR. Reviewing this will already be challenging enough without having to sift through hundreds of unrelated changes. To highlight the problem. Github won't even show me all of the changed files because this PR exceeds 300 changed files.

Please trim down the PR to only contain things related to LibGodot in order to ease the burden on PR reviewers and ensure we can review this in a reasonable amount of time. Keep in mind that most of us are doing this in our free time, so anything you can do to reduce the burden will have a direct impact on how soon this can get properly reviewed

I have left a few comments after a preliminary review. There is still a lot of work to do.

zorbathut commented:
> compartmentalize this PR into the First minimum core changes, independent of platforms, to speed up the review process and to ensure that CI/CD GitHub action passing in agile way.

That's what #110441 is :)

I'd be interested in @kisg's opinion on [my modifications to libgodot_create_godot_instance](https://github.com/godotengine/godot/pull/110441/commits/caf6f8120002c785d68cabb48c609cad696f9551#diff-00b0d880415f34e10cb797bc87557f3e1ebce00b2b8773c487dc954aff75c324R46-R64); instead of adding extra callbacks on every version I've chosen to include the necessary stuff (argc, argv, gdextension callback) and then basically a C-api-esque key/value dictionary for "further parameters". Which will hopefully be more long-term stable than a fixed DLL function.

I think that's the only significant change I made though, the rest is just all streamlining and removing stuff that isn't the core.

That said, I've spent roughly no effort trying to grok the embedded-display stuff; it's not a feature I need, I just want to get the barebones in place. I *think* this API will be able to reimplement the embedded-display stuff without signature changes, but a second opinion would be good.

kisg commented:
Dear @clayjohn,

thank you very much for this early review. We are of course happy to break this up into smaller PRs, which we also offered back when we first submitted this PR.

This code drop was rushed out mostly to show what is in the pipeline (in reference to the other PR), and we even had to leave out some other pieces, like unifying the GL Manager code across all platforms (including EGL, ANGLE and platform specific GL-host integrations), which we also were using successfully in production with LibGodot.

Kind regards,
Gergely

kisg commented:
> I'd be interested in @kisg's opinion on [my modifications to libgodot_create_godot_instance](https://github.com/godotengine/godot/pull/110441/commits/caf6f8120002c785d68cabb48c609cad696f9551#diff-00b0d880415f34e10cb797bc87557f3e1ebce00b2b8773c487dc954aff75c324R46-R64); instead of adding extra callbacks on every version I've chosen to include the necessary stuff (argc, argv, gdextension callback) and then basically a C-api-esque key/value dictionary for "further parameters". Which will hopefully be more long-term stable than a fixed DLL function.
> 

I have a solution in the works that integrates better with Godot Core, reduces the maintenance burden going forward, and does not require additional C-style functions in the future. I will be ready to show it before the Core meeting next week - I already added the topic to the meeting's agenda. It would be great if we could discuss it there if you have the time: https://files.godot.foundation/s/7jBFMWXjPg6EDsk?dir=/&openfile=true

zorbathut commented:
> I have a solution in the works that integrates better with Godot Core, reduces the maintenance burden going forward, and does not require additional C-style functions in the future. I will be ready to show it before the Core meeting next week - I already added the topic to the meeting's agenda. It would be great if we could discuss it there if you have the time: https://files.godot.foundation/s/7jBFMWXjPg6EDsk?dir=/&openfile=true

Could you post a summary of it? I'd love to see a better idea!

EricApostal commented:
> DisplayServer embedding is being used in production on Windows and iOS, Mac should work as well, but we don't have a sample for that yet. 

Are you sure this PR is the same as what you run in production? I'm still stuck on getting the DisplayServer running, and it seems as if it's disabled on anything other than MacOS. [This line](https://github.com/migeran/godot/blob/b4bc1a76702f56f0dcf789ace36180e0b96ca2b3/main/main.cpp#L1413) blocks DisplayServer from working at all on platforms other than MacOS, which was indeed hit when I tried to run in embedded display mode on Windows.

@kisg

kisg commented:
@EricApostal good catch, this must have crept in with one of the rebases. We will do a fix for this.

emmauss commented:
Being able to render to an external opengl texture or vulkan image is already a great improvement, especially for embedding into existing contexts that are managed by the external application.

AThousandShips commented on editor/audio/audio_stream_preview.cpp:
```diff
@@ -253,3 +253,7 @@ AudioStreamPreviewGenerator::AudioStreamPreviewGenerator() {
 }
+
+AudioStreamPreviewGenerator::~AudioStreamPreviewGenerator() {
+	singleton = nullptr;
```
Why no check here?

AThousandShips commented on platform/android/rendering_context_driver_vulkan_android.cpp:
```diff
@@ -29,6 +29,8 @@
 /**************************************************************************/
 
 #include "rendering_context_driver_vulkan_android.h"
+#include "drivers/vulkan/rendering_native_surface_vulkan.h"
```
```suggestion

#include "drivers/vulkan/rendering_native_surface_vulkan.h"
```

AThousandShips commented on platform/android/rendering_native_surface_android.cpp:
```diff
@@ -0,0 +1,67 @@
+/**************************************************************************/
+
+#include "rendering_native_surface_android.h"
+#import "rendering_context_driver_vulkan_macos.h"
```
This is incorrect now no? This is no longer objective C

AThousandShips commented on platform/linuxbsd/wayland/rendering_native_surface_wayland.cpp:
```diff
@@ -0,0 +1,68 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "rendering_native_surface_wayland.h"
```
```suggestion
#include "rendering_native_surface_wayland.h"

```

AThousandShips commented on platform/linuxbsd/x11/rendering_native_surface_x11.h:
```diff
@@ -0,0 +1,63 @@
+		return display;
+	}
+
+	RenderingContextDriver *create_rendering_context(const String &p_driver_name) override;
```
```suggestion
	virtual RenderingContextDriver *create_rendering_context(const String &p_driver_name) override;
```

AThousandShips commented on platform/windows/rendering_native_surface_windows.h:
```diff
@@ -0,0 +1,70 @@
+		return (void *)window;
+	}
+
+	RenderingContextDriver *create_rendering_context(const String &p_driver_name) override;
```
```suggestion
	virtual RenderingContextDriver *create_rendering_context(const String &p_driver_name) override;
```

GeorgeS2019 commented:
@kisg I think the community is ready for Migeran 4.6 upstreaming :-) 

### GodotCon 2026 
[April 23, 2026]
| Description | Development Status | Upstream Status |
| :--- | :--- | :--- |
| Core LibGodot | Stable | Merged in Godot 4.6 |
| Reinitialization Support | Stable in Migeran 4.6 Branch | PR Ready for Upstreaming |
| UI Embedding Infrastructure | Stable in Migeran 4.6 Branch | PR Ready for Upstreaming |
| Apple UI Embedding Support | Stable in Migeran 4.6 Branch | PR Ready for Upstreaming |
| Android UI Embedding Support | Stable in Migeran 4.6 Branch | PR In Progress |
| Windows UI Embedding Support | Stable in Migeran 4.6 Branch | PR In Progress |
